pkg/edgehub/controller_test.go

```
@@ -0,0 +1,889 @@
+	"github.com/kubeedge/kubeedge/pkg/edgehub/config"
+)
+
+//TestServer is a fake http server created for testing
+var TestServer *httptest.Server
+
+// mockAdapter is a mocked adapter implementation
+var mockAdapter *edgehub.MockAdapter
+
+//init() starts the test server and generates test certificates for testing
+func init() {
+	newTestServer()
+	generateTestCertificate()
+}
+
+// initMocks is function to initialize mocks
+func initMocks(t *testing.T) {
+	mockCtrl := gomock.NewController(t)
+	defer mockCtrl.Finish()
+	mockAdapter = edgehub.NewMockAdapter(mockCtrl)
+}
+
+//newTestServer() starts a fake server for testing
+func newTestServer() {
+	flag := true
+	TestServer = httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		switch {
+		case strings.Contains(r.RequestURI, "/proper_request"):
+			w.Write([]byte("ws://127.0.0.1:20000"))
+		case strings.Contains(r.RequestURI, "/bad_request"):
+			w.WriteHeader(http.StatusBadRequest)
+		case strings.Contains(r.RequestURI, "/wrong_url"):
+			if flag {
+				w.WriteHeader(http.StatusNotFound)
+				flag = false
+			} else {
+				w.Write([]byte("ws://127.0.0.1:20000"))
+			}
+		}
+	}))
+}
+
+//generateTestCertificate() generates fake certificates and stores them at a temporary location (/tmp/kubeedge/testData)
+func generateTestCertificate() {
+	template := &x509.Certificate{
+		IsCA:                  true,
+		BasicConstraintsValid: true,
+		SubjectKeyId:          []byte{1, 2, 3},
+		SerialNumber:          big.NewInt(1234),
+		Subject: pkix.Name{
+			Country:      []string{"test"},
+			Organization: []string{"testor"},
+		},
+		NotBefore:   time.Now(),
+		NotAfter:    time.Now().AddDate(5, 5, 5),
+		ExtKeyUsage: []x509.ExtKeyUsage{x509.ExtKeyUsageClientAuth, x509.ExtKeyUsageServerAuth},
+		KeyUsage:    x509.KeyUsageDigitalSignature | x509.KeyUsageCertSign,
+	}
+
+	// generate private key
+	privateKey, err := rsa.GenerateKey(rand.Reader, 2048)
+	if err != nil {
+		fmt.Println(err)
```

Let this function return error to make sure creation of certificates was success or not rather than just printing error.